feat: Add MLQA dataset for CrossLingual Retrieval

imenelydiaker opened the pull request:
I have made an attempt to add MLQA cross lingual retrieval dataset. I've updated the `AbsTaskRetrieval` to handle cross lingual datasets. I ran the task on one language pair Arabic and German: the corpus is in Arabic and the questions in German, you can see the results in the uploaded json files for `intfloat/multilingual-e5-small`.

I'm still missing this task:

- [x] Better handle the language mapping and loading (I know how to do it, will do tomorrow)

<!-- If you are not submitting for a dataset, feel free to remove the content below  -->


<!-- add additional description, question etc. related to the new dataset -->

## Checklist for adding MMTEB dataset

<!-- 
Before you commit here is a checklist you should complete before submitting
if you are not 
 -->
Reason for dataset addition:
<!-- Add reason for adding dataset here. E.g. it covers task/language/domain previously not covered -->


- [x] I have tested that the dataset runs with the `mteb` package.
- [x] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb run -m {model_name} -t {task_name}` command.
  - [x] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [x] `intfloat/multilingual-e5-small`
- [x] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [x] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [x] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 
- [ ] I have added points for my submission to the [points folder](https://github.com/embeddings-benchmark/mteb/blob/main/docs/mmteb/points.md) using the PR number as the filename (e.g. `438.jsonl`).

imenelydiaker commented on mteb/tasks/Retrieval/multilingual/MLQARetrieval.py:
```diff
@@ -0,0 +1,162 @@
+        _dataset_raw = {}
+        self.queries, self.corpus, self.relevant_docs = {}, {}, {}
+
+        for lang in _LANGUAGES.keys():
```
This loop loads all language pairs, I'll update it to use `self.langs` and upload only specified language pairs

isaac-chung commented on mteb/tasks/Retrieval/multilingual/MLQARetrieval.py:
```diff
@@ -0,0 +1,162 @@
+    for x in _ISO6393_to_ISO6391.keys():
+        for y in _ISO6393_to_ISO6391.keys():
+            if x != y:
+                pair = f"{x}_{y}"
```
Would each pair be query in xx -> corpus in yy?

imenelydiaker replied:
The dataset has a subset for each language pair xx->yy, I'd prefer not mixing them to avoid errors.

So what I did for now is use each subset: xx->yy on it's own. Since all texts are machine translated. 

You can see the subsets in the _LANGUAGES variable. Although, I'll change the way I'm handling these pairs because this is not very clear to understand.

isaac-chung replied:
Oh yes I'll add to the discussion. I think whichever convention we end up choosing, we just need to write it down somewhere. I tend to forget 🙈 , so it'll be helpful to be able to check from time to time.

imenelydiaker commented:
@isaac-chung this PR is ready for review. I ran both models on one language pair, but a compelete evaluation is coming (takes some time for 49 pairs).

I didn't add the named tuple as discussed here, as it requires to update all CrossLingual tasks. I prefer adding this in another PR with all necessary refactorings.

For the now, the language is handled this way:
```py
eval_langs = {
"eng-Latn_fra-Latn": ["eng-Latn", "fra-Latn"],
...
}
```

We can call the task  on a language pair (e.g., from English to Spanish) like this:
```bash
mteb -t MLQARetrieval -m intfloat/multilingual-e5-small -l eng-Latn_spa-Latn
```

isaac-chung commented on mteb/tasks/Retrieval/multilingual/MLQARetrieval.py:
```diff
@@ -0,0 +1,152 @@
+        for hf_subset, langs in _LANGUAGES.items():
+            lang_pair = (
+                langs[0] + "_" + langs[1]
+            )  # builds a language pair separated by an underscore. e.g., "ara-Arab_eng-Latn"
```
What do you think about, instead of `eng-Latn_fra-Latn`, we only use the 3 letter lang code, i.e. `eng_fra` ? As they should be unique, it saves a lot of typing. 

Aside: could also consider making line 118 into its own small function as it's used more than once.

imenelydiaker replied:
I like this proposition, I'm afraid the format is not adequate to our language standard. I'll give it a try and let you know!

isaac-chung replied:
It should be fine. When passed into `-l`, MTEB calls `_extend_lang_code`, which does a mapping if only a 3 letter code is provided. So both standards should be supported without problems.

isaac-chung reviewed (commented):
I think the rest looks good. Just a few small comments. 

Separating the named tuple into its own PR sounds like a good plan 👍

imenelydiaker commented on mteb/abstasks/AbsTaskRetrieval.py:
```diff
@@ -246,7 +246,7 @@ def evaluate(self, model, split="test", **kwargs):
+        if self.is_crosslingual or self.is_multilingual:
             for lang in self.langs:
                 logger.info(f"Language: {lang}")
                 corpus, queries, relevant_docs = (
```
True, should I name it eval_split as in this task: [AbsTaskSTS](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/abstasks/AbsTaskSTS.py) ?

imenelydiaker commented on mteb/tasks/Retrieval/multilingual/MLQARetrieval.py:
```diff
@@ -0,0 +1,154 @@
+        year = 2019,
+        eid = {arXiv: 1910.07475}
+        }""",
+        n_samples={"test": 158083, "validation": 15747},
```
Yeah for now the number of samples for multilingual and crosslingual tasks is just equal to the sum of samples accross all languages and pairs. We're still thinking about how to implement a faster version of tasks for Retrieval, just as we did in [Clustering,](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/abstasks/AbsTaskClusteringFast.py) do you have any suggestions? We can't simply subsample imo, because it will impact the ndcg score..

Andrian0s replied:
I agree, downsampling is quite bad, and the large amount of samples issue itself in this pull request spawns from the amount of language pairs, which is natural and unavaoidable. 

A non trivial solution would be very nice, but I would hope for a proper reasoning defending that choice.

For now, I would keep as is (since the embeddings are currently not calculated multiple times, so all good). I would still suggest to avoid the validation set, test set is enough.

imenelydiaker commented:
@isaac-chung @Andrian0s I fixed the code to build language pair in this format: `lang1-lang2` (see screenshot below), and renamed `_eval_monolingual()` in `AbsTaskRetrieval`.I think it's ready for merging 🙂 
![image](https://github.com/embeddings-benchmark/mteb/assets/33312980/9875b0ff-2547-48c4-9d4b-796ddcd03759)

Andrian0s commented:
> @isaac-chung @Andrian0s I fixed the code to build language pair in this format: `lang1-lang2` (see screenshot below), and renamed `_eval_monolingual()` in `AbsTaskRetrieval`.I think it's ready for merging 🙂
> ![image](https://private-user-images.githubusercontent.com/33312980/327966558-9875b0ff-2547-48c4-9d4b-796ddcd03759.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ4NDkwMDgsIm5iZiI6MTcxNDg0ODcwOCwicGF0aCI6Ii8zMzMxMjk4MC8zMjc5NjY1NTgtOTg3NWIwZmYtMjU0Ny00OGM0LTlkNGItNzk2ZGRjZDAzNzU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA0VDE4NTE0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhMWRhN2Q1NmM0YTFiYzMzMjNhZTYzY2MyNTg0NWNhZDk0YmU2ZjRjNzk2NDUzMjAwZDFjNDg0MWRkYjExODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.UZ2cznlfTKQYU0TdrNu_p-MnbvcXOlIHVIkP0k-DU2w)

Looks good. If the tests pass, I would also support merging.

imenelydiaker commented:
Okay I may have broken something, tests are all good locally.. I'm checking

isaac-chung reviewed (approved):
Thanks for adding this Imene!
